src/serlio/materials/StingrayMaterialNode.cpp

```diff
@@ -0,0 +1,236 @@
+/**
+ * Serlio - Esri CityEngine Plugin for Autodesk Maya
+ *
+ * See https://github.com/esri/serlio for build and usage instructions.
+ *
+ * Copyright (c) 2012-2019 Esri R&D Center Zurich
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "materials/StingrayMaterialNode.h"
+#include "materials/MaterialInfo.h"
+#include "materials/MaterialUtils.h"
+
+#include "modifiers/PRTModifierAction.h"
+
+#include "utils/MELScriptBuilder.h"
+#include "utils/MItDependencyNodesWrapper.h"
+#include "utils/MayaUtilities.h"
+
+#include "PRTContext.h"
+#include "serlioPlugin.h"
+
+#include "maya/MFnMesh.h"
+#include "maya/MFnTypedAttribute.h"
+#include "maya/MGlobal.h"
+#include "maya/MItDependencyNodes.h"
+#include "maya/MObject.h"
+#include "maya/adskDataAssociations.h"
+#include "maya/adskDataStream.h"
+
+#include <array>
+#include <mutex>
+#include <set>
+
+namespace {
+
+constexpr bool DBG = false;
+
+const std::wstring MATERIAL_BASE_NAME = L"serlioStingrayMaterial";
+
+std::once_flag pluginDependencyCheckFlag;
+const std::vector<std::string> PLUGIN_DEPENDENCIES = {"shaderFXPlugin"};
+
+void setTexture(MELScriptBuilder& sb, const MELVariable& shaderNode, const std::wstring& target,
+                const std::wstring& textureNodeBaseName, const std::wstring& tex) {
+	if (!tex.empty()) {
+		MELVariable mapNode(L"mapNode");
+		sb.setVar(mapNode, textureNodeBaseName);
+
+		MELVariable mapFile(L"mapFile");
+		sb.setVar(mapFile, tex);
+
+		sb.createTextureShadingNode(mapNode);
+		sb.setAttr(mapNode, L"fileTextureName", mapFile.mel());
+
+		sb.connectAttr(mapNode, L"outColor", shaderNode, L"TEX_" + target);
+		sb.setAttr(shaderNode, L"use_" + target, 1);
+	}
+	else {
+		sb.setAttr(shaderNode, L"use_" + target, 0);
+	}
+}
+
+void appendToMaterialScriptBuilder(MELScriptBuilder& sb, const MaterialInfo& matInfo,
+                                   const std::wstring& shaderBaseName, const std::wstring& shadingEngineName) {
+	// create shader
+	MELVariable shaderNode(L"shaderNode");
+	sb.setVar(shaderNode, shaderBaseName);
+	sb.setVar(MEL_VARIABLE_SHADING_ENGINE, shadingEngineName);
+	sb.createShader(L"StingrayPBS", shaderNode);
+
+	// connect to shading group
+	sb.connectAttr(shaderNode, L"outColor", MEL_VARIABLE_SHADING_ENGINE, L"surfaceShader");
+
+	// stingray specifics
+	const std::wstring sfxFile = MaterialUtils::getStingrayShaderPath();
+	sb.addCmdLine(L"shaderfx -sfxnode $shaderNode -loadGraph  \"" + sfxFile + L"\";");

```

Comment:
consider using `shaderNode.mel()` instead of `$shaderNode`

Reply:
done